src/navigation/drive.py

```diff
@@ -1,63 +1,323 @@
+        """
+        if self.path and target_pos == self.target_pos:
+            return self.path[self.target_vertex_idx]
+        else:
+            self.generate_path(source_pos, target_pos)
+
+
+    def complete_intermediate_target(self) -> None:
+        """
+        Marks the previously-returned intermediate target as complete. 
+        """  
+        assert(self.path) 
+
+        if self.target_vertex_idx < len(self.path):
+            self.target_vertex_idx += 1
+
+
+    def is_path_complete(self) -> bool:
+        """
+        Returns whether the path has been fully traversed, which happens
+        if and only if the target_pos has been marked as completed. 
+        """
+        return (self.path != None) and (self.target_vertex_idx == len(self.path))
+
+
+    def is_edge_safe(self, edge: LineString) -> bool:
+        """
+        Returns whether a given proposed edge is safe. An edge is deemed safe
+        if and only if it does not intersect with any failure zones. 
+
+        :param edge:    the edge to be checked for safety 
+        """
+        for fz in self.failure_zones:
+            if fz.intersects(edge):
+                return False
+        
+        return True
+
+
+    def add_failure_zone(self, failure_zone: FailureZone) -> None:
+        """
+        Adds a failure zone to the visibility graph 
+
+        :param failure_zone:    the failure zone to be added (as a FailureZone object)
+        """
+        # Step 1: add to list of failure zones
+        self.failure_zones.append(failure_zone)
+
+        # Step 2: remove all edges from current graph that cross this zone
+        graph = self.visibility_graph.get_data()
+        deleted_edges = []
+
+        for u, neighbors in graph.items():
+            for v, edge in neighbors.items():
+                if failure_zone.intersects(edge):
+                    deleted_edges.append((u, v))
+        
+        for (u, v) in deleted_edges:
+            self.visibility_graph.remove_edge(u, v)
+
+        # Step 3: Add corners to graph
+        vertices = failure_zone.get_vertices()
+        for vertex in vertices:
+            self.add_vertex(vertex)
+
+        # Step 4: Reset path, target, etc. to None
+        self.path = None
+        self.target_vertex_idx = 0
+
+
+    def add_vertex(self, new_vertex: Point):
+        """
+        Adds the given vertex to the visibility graph, along with all 
+        associated safe edges, and resets the path. 
+
+        :param new_vertex:  shapely Point object representing the vertex to be added
+        """
+        # Add vertex to graph
+        self.visibility_graph.add_node(new_vertex)
+
+        # Iterate through all vertices; if edge doesn't intersect any FZ, add to graph
+        for vertex in self.visibility_graph:
+            if(vertex != new_vertex):
+                edge = LineString(vertex, new_vertex)
+                if self.is_edge_safe(edge):
+                    self.visibility_graph.add_edge(vertex, new_vertex, edge)
```

Comment:
can you add a LineString object to this graph? Or do you want to get the length.

Reply:
adding a LineString to the graph is legal, we just use the length of the edge as the cost function when calculating the path. There's no reason currently that we need the LineString to be added (as opposed to just the length), but adding the full object makes the shortest path algorithm also return the list of LineStrings that need to be followed, which might be nice in the future. LineString objects are pretty light, so I don't think it should be much of an issue to keep it like this